prov/gni/src/gnix_atomic.c

```diff
@@ -0,0 +1,606 @@
+	return gnixu_to_fi_errno(status);
+}
+
+static int __gnix_amo_txd_complete(void *arg, gni_return_t tx_status)
+{
+	struct gnix_tx_descriptor *txd = (struct gnix_tx_descriptor *)arg;
+	struct gnix_fab_req *req = txd->req;
+	int rc = FI_SUCCESS;
+	uint32_t read_data32;
+	uint64_t read_data64;
+
+	if (tx_status != GNI_RC_SUCCESS) {
+		return __gnix_amo_post_err(txd);
+	}
+
+	/* FI_ATOMIC_READ data is delivered to operand buffer in addition to
+	 * the results buffer. */
+	if (req->amo.op == FI_ATOMIC_READ) {
+		switch(fi_datatype_size(req->amo.datatype)) {
+		case sizeof(uint32_t):
+			read_data32 = *(uint32_t *)req->amo.loc_addr;
+			*(uint32_t *)req->amo.read_buf = read_data32;
+			break;
+		case sizeof(uint64_t):
+			read_data64 = *(uint64_t *)req->amo.loc_addr;
+			*(uint64_t *)req->amo.read_buf = read_data64;
+			break;
+		default:
+			GNIX_WARN(FI_LOG_EP_DATA, "Invalid datatype: %d\n",
+				  req->amo.datatype);
+			assert(0);
+			break;
+		}
+	}
+
+	/* Successful delivery.  Progress request. */
+	if (req->flags & FI_REMOTE_CQ_DATA) {
+		/* initiate immediate data transfer */
+		req->tx_failures = 0;
+		req->txd = (void *)txd;
+		req->send_fn = __gnix_amo_send_data_req;
+		_gnix_vc_queue_req(req);
+	} else {
+		/* complete request */
+		rc = __gnix_amo_send_completion(req->vc->ep, req);
+		if (rc != FI_SUCCESS)
+			GNIX_WARN(FI_LOG_EP_DATA,
+				  "__gnix_amo_send_completion() failed: %d\n",
+				  rc);
+
+		__gnix_amo_fr_complete(req, txd);
+	}
+
+	return FI_SUCCESS;
+}
+
+/*
+ * Datatypes:
+ *
+ * FI_INT8, FI_UINT8, FI_INT16, FI_UINT16,
+ * FI_INT32, FI_UINT32,
+ * FI_INT64, FI_UINT64,
+ * FI_FLOAT, FI_DOUBLE,
+ * FI_FLOAT_COMPLEX, FI_DOUBLE_COMPLEX,
+ * FI_LONG_DOUBLE, FI_LONG_DOUBLE_COMPLEX
+ */
+
+static int __gnix_amo_cmds[FI_ATOMIC_OP_LAST][FI_DATATYPE_LAST] = {
+	/*
+	 * Basic AMO types:
+	 * FI_MIN, FI_MAX, FI_SUM, FI_PROD, FI_LOR, FI_LAND,  FI_BOR,  FI_BAND,
+	 *       FI_LXOR, FI_BXOR, and FI_ATOMIC_WRITE.
+	 */
+	[FI_MIN] = { 0,0,0,0, GNI_FMA_ATOMIC2_IMIN_S, 0, GNI_FMA_ATOMIC2_IMIN, 0, GNI_FMA_ATOMIC2_FPMIN_S, GNI_FMA_ATOMIC2_FPMIN },
+	[FI_MAX] = { 0,0,0,0, GNI_FMA_ATOMIC2_IMAX_S, 0, GNI_FMA_ATOMIC2_IMAX, 0, GNI_FMA_ATOMIC2_FPMAX_S, GNI_FMA_ATOMIC2_FPMAX },
+	[FI_SUM] = { 0,0,0,0, GNI_FMA_ATOMIC2_IADD_S, GNI_FMA_ATOMIC2_IADD_S, GNI_FMA_ATOMIC2_IADD, GNI_FMA_ATOMIC2_IADD, GNI_FMA_ATOMIC2_FPADD_S, 0 /* DP addition is broken */ },
+	[FI_BOR] = { 0,0,0,0, GNI_FMA_ATOMIC2_OR_S, GNI_FMA_ATOMIC2_OR_S, GNI_FMA_ATOMIC2_OR, GNI_FMA_ATOMIC2_OR, 0, 0 },
+	[FI_BAND] = { 0,0,0,0, GNI_FMA_ATOMIC2_AND_S, GNI_FMA_ATOMIC2_AND_S, GNI_FMA_ATOMIC2_AND, GNI_FMA_ATOMIC2_AND, 0, 0 },
+	[FI_BXOR] = { 0,0,0,0, GNI_FMA_ATOMIC2_XOR_S, GNI_FMA_ATOMIC2_XOR_S, GNI_FMA_ATOMIC2_XOR, GNI_FMA_ATOMIC2_XOR, 0, 0 },
+	[FI_ATOMIC_WRITE] = { 0,0,0,0, GNI_FMA_ATOMIC2_SWAP_S, GNI_FMA_ATOMIC2_SWAP_S, GNI_FMA_ATOMIC2_SWAP, GNI_FMA_ATOMIC2_SWAP, GNI_FMA_ATOMIC2_SWAP_S, GNI_FMA_ATOMIC2_SWAP },
+};
+
+static int __gnix_fetch_amo_cmds[FI_ATOMIC_OP_LAST][FI_DATATYPE_LAST] = {
```

Comment:
I should have been more explicit in my comment. This would be a good place
for a comment to describe what it is that the table is being used for. It
may be apparent to the designer, but the designer and the maintainer may
be two different people.




On 11/6/15, 3:34 PM, "Zach Tiffany" <notifications@github.com> wrote:

>In 
>prov/gni/src/gnix_atomic.c
><https://github.com/ofi-cray/libfabric-cray/pull/490#discussion_r44190116>
>:
>> +static int __gnix_amo_cmds[FI_ATOMIC_OP_LAST][FI_DATATYPE_LAST] = {
>> +	/*
>> +	 * Basic AMO types:
>> +	 * FI_MIN, FI_MAX, FI_SUM, FI_PROD, FI_LOR, FI_LAND,  FI_BOR,
>>FI_BAND,
>> +	 *       FI_LXOR, FI_BXOR, and FI_ATOMIC_WRITE.
>> +	 */
>> +	[FI_MIN] = { 0,0,0,0, GNI_FMA_ATOMIC2_IMIN_S, 0,
>>GNI_FMA_ATOMIC2_IMIN, 0, GNI_FMA_ATOMIC2_FPMIN_S, GNI_FMA_ATOMIC2_FPMIN
>>},
>> +	[FI_MAX] = { 0,0,0,0, GNI_FMA_ATOMIC2_IMAX_S, 0,
>>GNI_FMA_ATOMIC2_IMAX, 0, GNI_FMA_ATOMIC2_FPMAX_S, GNI_FMA_ATOMIC2_FPMAX
>>},
>> +	[FI_SUM] = { 0,0,0,0, GNI_FMA_ATOMIC2_IADD_S, GNI_FMA_ATOMIC2_IADD_S,
>>GNI_FMA_ATOMIC2_IADD, GNI_FMA_ATOMIC2_IADD, GNI_FMA_ATOMIC2_FPADD_S, 0
>>/* DP addition is broken */ },
>> +	[FI_BOR] = { 0,0,0,0, GNI_FMA_ATOMIC2_OR_S, GNI_FMA_ATOMIC2_OR_S,
>>GNI_FMA_ATOMIC2_OR, GNI_FMA_ATOMIC2_OR, 0, 0 },
>> +	[FI_BAND] = { 0,0,0,0, GNI_FMA_ATOMIC2_AND_S, GNI_FMA_ATOMIC2_AND_S,
>>GNI_FMA_ATOMIC2_AND, GNI_FMA_ATOMIC2_AND, 0, 0 },
>> +	[FI_BXOR] = { 0,0,0,0, GNI_FMA_ATOMIC2_XOR_S, GNI_FMA_ATOMIC2_XOR_S,
>>GNI_FMA_ATOMIC2_XOR, GNI_FMA_ATOMIC2_XOR, 0, 0 },
>> +	[FI_ATOMIC_WRITE] = { 0,0,0,0, GNI_FMA_ATOMIC2_SWAP_S,
>>GNI_FMA_ATOMIC2_SWAP_S, GNI_FMA_ATOMIC2_SWAP, GNI_FMA_ATOMIC2_SWAP,
>>GNI_FMA_ATOMIC2_SWAP_S, GNI_FMA_ATOMIC2_SWAP },
>> +};
>> +
>> +static int __gnix_fetch_amo_cmds[FI_ATOMIC_OP_LAST][FI_DATATYPE_LAST]
>>= {
>It's a lookup table.
>(amo operation, datatype) -> GNI amo command
>—
>Reply to this email directly or
>view it on GitHub 
><https://github.com/ofi-cray/libfabric-cray/pull/490/files#r44190116>.
>
>